Add new badges to Audience Selection Panel

nfmohit opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses in the following list. -->
Addresses issue:

- #8170

## Relevant technical choices

- This PR adds new badges to audiences in the Audience Selection Panel.

### Deviations from IB

> * Create a `useEffect` in this component which calls the `setExpirableItemTimers` action, _only if the user has never seen New badge before_, and the title is not either of the pre-defined audiences , with the following object:
>   ```
>   {
>      `audience-selection-panel-expirable-new-badge-${slug}`: WEEK_IN_SECONDS * 4
>   }
>   ```

Instead of doing the above, this PR sets the expiry for all new badges at once in the `AudienceItems` level. This way, we reduce the number of HTTP requests used to make this happen to one.

> * Update [`assets/js/modules/analytics-4/datastore/audiences.js`](https://github.com/google/site-kit-wp/blob/51452f6f0d6e3afedfbfc841c5649d328be6ca0f/assets/js/modules/analytics-4/datastore/audiences.js#L346-L356), once the `configuredAudiences` are successfully saved, dispatch an `setExpirableItemTimers` action for every `configuredAudiences` using the slug, and passing a value of 0, which will expire these items immediately.

A value of `0` is not accepted by the `expirable-items` infrastructure, so we're settings this to `1`.

> ### Test Coverage
> * Update Storybook to include cases with a new badge present for both the generic Selection Panel and the Audience Selection Panel.

In addition to the above, for better test coverage, we've also added coverage for the badge visibility in the Audience Selection Panel and the family of `enable*AudienceGroup*` actions.

## PR Author Checklist

- [ ] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 5.2 and PHP 7.4.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [x] Run the code.
- [x] Ensure the acceptance criteria are satisfied.
- [ ] Reassess the implementation with the IB.
- [x] Ensure no unrelated changes are included.
- [ ] Ensure CI checks pass.
- [x] Check Storybook where applicable.
- [x] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [x] Ensure the PR has the correct target branch.
- [x] Double-check that the PR is okay to be merged.
- [x] Ensure the corresponding issue has a ZenHub release assigned.
- [x] Add a changelog message to the issue.

nfmohit commented on assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSelectionPanel/AudienceItems.js:
```diff
@@ -214,6 +220,49 @@ export default function AudienceItems( { savedItemSlugs = [] } ) {
+	// cause the effect to unexpectedly run multiple times.
+	useDeepCompareEffect( () => {
+		if ( newBadgesToActivate !== undefined && newBadgesToActivate.length ) {
+			setExpirableItemTimers(
+				newBadgesToActivate.map( ( slug ) => ( {
+					slug,
+					expiresInSeconds: WEEK_IN_SECONDS * 4,
+				} ) )
+			);
```
Setting expiry for the badges here instead of the individual audience items so that we don't have to make one HTTP request for each audience.

nfmohit commented on assets/js/modules/analytics-4/datastore/audiences.js:
```diff
@@ -522,6 +523,16 @@ const baseActions = {
+			dispatch( CORE_USER ).setExpirableItemTimers(
+				configuredAudiences.map( ( slug ) => ( {
+					slug: `${ AUDIENCE_ITEM_NEW_BADGE_SLUG_PREFIX }${ slug }`,
+					expiresInSeconds: 1,
```
The `expirable-items` infrastructure does not accept `0` as a value for `expiresInSeconds`, so we're using `1` so that it expires the new badges as soon as possible, i.e. 1 second.

nfmohit commented on assets/js/modules/analytics-4/datastore/audiences.js:
```diff
@@ -619,6 +630,16 @@ const baseActions = {
+			dispatch( CORE_USER ).setExpirableItemTimers(
+				configuredAudiences.map( ( slug ) => ( {
+					slug: `${ AUDIENCE_ITEM_NEW_BADGE_SLUG_PREFIX }${ slug }`,
+					expiresInSeconds: 1,
```
Same as above.

nfmohit commented on tests/backstop/reference/google-site-kit_Modules_Analytics4_Components_AudienceSegmentation_Dashboard_AudienceSelectionPanel_AudienceCreationNoticeOneAdded_0_document_0_small.png:
This and other VRT changes (not additions) are due to the change in the spacing between the audience label and the sub-title in the Figma designs.

hussain-t commented on assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSelectionPanel/AudienceItem.js:
```diff
@@ -36,23 +36,37 @@ import {
+import NewBadge from '../../../../../../components/NewBadge';
 import { SelectionPanelItem } from '../../../../../../components/SelectionPanel';
 
+export const AUDIENCE_ITEM_NEW_BADGE_SLUG_PREFIX =
+	'audience-selection-panel-expirable-new-badge-';
```
Since it's being used elsewhere, including the datastore, can we create it in `assets/js/modules/analytics-4/datastore/constants.js`?

hussain-t commented on assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSelectionPanel/index.test.js:
```diff
@@ -16,15 +16,22 @@
  * limitations under the License.
  */
 
+/**
+ * External dependencies
+ */
+import fetchMock from 'fetch-mock';
+
```
We don't need to import `fetchMock`. It's globally available.
```suggestion
```

hussain-t commented on assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceSelectionPanel/index.stories.js:
```diff
@@ -284,6 +285,17 @@ AudienceCreationSuccessNotice.scenario = {
+		registry.dispatch( CORE_USER ).receiveGetExpirableItems( {} );
+	},
+};
+WithNewBadges.scenario = {
+	label: 'Modules/Analytics4/Components/AudienceSegmentation/Dashboard/AudienceSelectionPanel/WithNewBadges',
+};
```
We don't need to provide the label. It will be generated.
```suggestion
WithNewBadges.scenario = {};
```

hussain-t reviewed (changes requested):
Great work, @nfmohit. The changes look solid overall. I’ve added a few minor comments for your review.

In addition, could you update the QAB to include a scenario where the “New” badge appears after Site Kit audiences are archived in the Analytics console and then re-created through the Audience Creation Notice?

nfmohit commented:
> Great work, @nfmohit. The changes look solid overall. I’ve added a few minor comments for your review.
> 
> In addition, could you update the QAB to include a scenario where the “New” badge appears after Site Kit audiences are archived in the Analytics console and then re-created through the Audience Creation Notice?

I've addressed the CR feedback, and have updated the QAB as requested. Thank you @hussain-t !

hussain-t reviewed (changes requested):
Thanks for addressing the feedback, @nfmohit. The `With_New_badges` VRT is failing and seems like missing reference images. Could you download the `vrt-report` and check it?

nfmohit commented:
> Thanks for addressing the feedback, @nfmohit. The `With_New_badges` VRT is failing and seems like missing reference images. Could you download the `vrt-report` and check it?

Thank you @hussain-t, addressed.

hussain-t reviewed (approved):
Excellent work, @nfmohit. LGTM 👍
